Mark OpenSSLDie noreturn, supersedes #525

hesiod opened the pull request:
This is the revised version of PR #525.

It is good practice to mark functions that don't return as `noreturn`/`_Noreturn`. The former is a definition found in `stdnoreturn.h`, the latter is the actual keyword. Both are available from C11 onwards. If C11 is not used, optionally fall back to the GCC-specific `__attribute__((noreturn))` or to nothing at all if the compiler is not GCC.

This change particularly helps in reducing static analysis false positives.

richsalz commented:
@levitte, should this be ```__onoreturn``` like ```__owur```?  Your call, but I will +1 this if you want to take on the merge.

hesiod commented:
@richsalz, I'm obviously not a veteran OpenSSL contributor, but noreturn is not only a 'regular' attribute, but more of a core element of the C(11) language. The only reason of having the 'workaround' with `stdnoreturn.h` and `_Noreturn` is backwards compatibility, I guess.

levitte commented:
@richsalz makes a good point.  Our redefining `noreturn` might disturb others having the same idea, so it would be preferable to do like...  I think I'd prefer that to be like `ossl_inline`, so @hesiod, if you could make it so `e_os2.h` defines `ossl_noreturn` and uses that, I'd be quite happy with this change.

BTW, you don't have to make a new pull request just to change a commit.  Just make a fixup commit, then `git rebase -i` with appropriate arguments, and `git push -f` when done.

richsalz commented:
And if that change is made, then my +1 still stands and @levitte can do the merge when he gets a chance.

hesiod commented:
@levitte, sometimes, when I'm in a lazy mood, I just rely on the point-and-click Github GUI.

Anyways, will do the changes.

levitte commented on include/openssl/e_os2.h:
```diff
@@ -350,6 +350,16 @@ typedef unsigned __int64 uint64_t;
+# ifdef __STDC_VERSION__
+#  if (__STDC_VERSION__ >= 201112L)
+#   include <stdnoreturn.h>
+#  endif
```
There's a logic flaw here that will mean that for any `__STDC_VERSION__` lower than 201112L, there will be no `noreturn`

hesiod replied:
You're right - we've opened the box of pandora or in this case, the box of compiler feature detection macros...

levitte replied:
Let's not exagerate ;-) 
I'd think the first condition should be like this :

    #if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L

...unless I'm missing something.

hesiod commented:
I fixed the logic flaw and renamed noreturn to ossl_noreturn.

Note that by defining noreturn ourselves when using C11, we do not rely on `stdnoreturn.h` and any compatibility logic contained there, but afaik no platform does anything fancy in `stdnoreturn.h` at the moment.

richsalz commented:
ping @levitte  -- you okay with this?  I'll merge if you plus-one.

levitte commented on include/openssl/e_os2.h:
```diff
@@ -350,6 +350,14 @@ typedef unsigned __int64 uint64_t;
 
+# if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L
+#  define ossl_noreturn _Noreturn
+# elif defined(__GNUC__)
```
You might need to figure out which GCC version that started supporting this attribute and adapt the condition accordingly.

hesiod replied:
Do you mean the C11-compliant `_Noreturn` above or the `__attribute__((noreturn))` below? The former [appeared in GCC 4.7](https://gcc.gnu.org/gcc-4.7/changes.html), and I'm not quite sure about the latter.

I guess the best option would be to order the directives so that we first test for GCC < 4.7 and then for C11.

mattcaswell commented:
Ping @levitte

levitte commented:
This looks much better.  +1

richsalz commented:
merged.
thanks for your patience!
